pkg/cmd/cli/cmd/loginoptions.go

```diff
@@ -71,122 +36,178 @@ func (o *LoginOptions) GatherInfo() error {
-
-		if clientcmd.IsCertificateAuthorityUnknown(result.Error()) {
-			fmt.Println("The server uses a certificate signed by an unknown authority.")
-			fmt.Println("You can bypass the certificate check, but any data you send to the server could be intercepted by others.")
+		switch {
+		case o.InsecureTLS:
+			clientConfig.Insecure = true
 
-			clientCfg.Insecure = util.PromptForBool(os.Stdin, "Use insecure connections? (y/n): ")
-			if !clientCfg.Insecure {
-				return fmt.Errorf(clientcmd.GetPrettyMessageFor(result.Error()))
+		// certificate issue, prompt user for insecure connection
+		case clientcmd.IsCertificateAuthorityUnknown(result.Error()):
+			// check to see if we already have a cluster stanza that tells us to use --insecure for this particular server.  If we don't, then prompt
+			clientConfigToTest := *clientConfig
+			clientConfigToTest.Insecure = true
+			matchingClusters := getMatchingClusters(clientConfigToTest, *o.StartingKubeConfig)
+
+			if len(matchingClusters) > 0 {
+				clientConfig.Insecure = true
+
+			} else if cmdutil.IsTerminal(o.Reader) {
+				fmt.Println("The server uses a certificate signed by an unknown authority.")
+				fmt.Println("You can bypass the certificate check, but any data you send to the server could be intercepted by others.")
+
+				clientConfig.Insecure = cmdutil.PromptForBool(os.Stdin, "Use insecure connections? (y/n): ")
+				if !clientConfig.Insecure {
+					return nil, fmt.Errorf(clientcmd.GetPrettyMessageFor(result.Error()))
+				}
+				fmt.Println()
 			}
-			fmt.Println()
-		} else {
-			return result.Error()
+
+		default:
+			return nil, result.Error()
 		}
 	}
 
-	// we have all info we need, now we can have a proper Config
+	o.Config = clientConfig
+	return o.Config, nil
+}
 
-	o.Config = clientCfg
+// getMatchingClusters examines the kubeconfig for all clusters that point to the same server
+func getMatchingClusters(clientConfig kclient.Config, kubeconfig clientcmdapi.Config) util.StringSet {
+	ret := util.StringSet{}
 
-	o.gatheredServerInfo = true
-	return nil
+	for key, cluster := range kubeconfig.Clusters {
+		if (cluster.Server == clientConfig.Host) && (cluster.InsecureSkipTLSVerify == clientConfig.Insecure) && (cluster.CertificateAuthority == clientConfig.CAFile) && (bytes.Compare(cluster.CertificateAuthorityData, clientConfig.CAData) == 0) {
+			ret.Insert(key)
+		}
+	}
+
+	return ret
 }
 
 // Negotiate a bearer token with the auth server, or try to reuse one based on the
 // information already present. In case of any missing information, ask for user input
 // (usually username and password, interactive depending on the Reader).
 func (o *LoginOptions) gatherAuthInfo() error {
-	if err := o.assertGatheredServerInfo(); err != nil {
+	directClientConfig, err := o.getClientConfig()
+	if err != nil {
 		return err
 	}
-
-	if me, err := o.whoAmI(); err == nil && (!o.usernameProvided() || (o.usernameProvided() && o.Username == me.Name)) {
```

Comment:
----- Original Message -----
> >  		return err
> >  	}
> > -
> > -	if me, err := o.whoAmI(); err == nil && (!o.usernameProvided() ||
> > (o.usernameProvided() && o.Username == me.Name)) {
> 
> > Meaning that, two consecutive runs of osc login (without flags) will always
> > ask for the username, even if the token in the users entry that is part of
> > the current-context is still valid? (I'm not against this approach, just
> > making sure I understood).
> 
> Yes, that's what this does.  Since stanzas are re-used and tokens are
> searched for matches against the same server, it won't muddy the kubeconfig
> file and it keeps the command doing the same thing.
> 
> It might be nice to build a "whoami" designed for readable output about you.
> Things like your name, which projects you have in your kubeconfig, and the
> top ten other projects you can view.

osc project -v maybe.  Agree though we need it.  We need to add it to the list.